### PR TITLE
[SPARK-36741][SQL] ArrayDistinct handle duplicated Double.NaN and Float.Nan

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
For query
```
select array_distinct(array(cast('nan' as double), cast('nan' as double)))
```
This returns [NaN, NaN], but it should return [NaN].
This issue is caused by `OpenHashSet` can't handle `Double.NaN` and `Float.NaN` too.
In this pr fix this based on https://github.com/apache/spark/pull/33955

### Why are the changes needed?
Fix bug

### Does this PR introduce _any_ user-facing change?
ArrayDistinct won't show duplicated `NaN` value

### How was this patch tested?
Added UT

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 1]
```diff
@@ -2326,4 +2326,13 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
       Literal.create(Seq(Float.NaN, null, 1f), ArrayType(FloatType))),
       Seq(Float.NaN, null, 1f))
   }
+
+  test("SPARK-36740: ArrayDistinct should handle duplicated Double.NaN and Float.Nan") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2330]: SPARK-36741 instead of SPARK-36740?

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala]: Done...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -3410,32 +3410,63 @@ case class ArrayDistinct(child: Expression)
...
+        if (array.isNullAt(i)) {
+          if (!hs.containsNull) {
+            hs.addNull
+            arrayBuffer += null
+          }
+        } else {
+          val elem = array.get(i, elementType)
+          if (isNaN(elem)) {
+            if (!hs.containsNaN) {
+              arrayBuffer += elem
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3433]: For this, let's wait for the decision at the first PR.
- https://github.com/apache/spark/pull/33955/files#r708570515

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: > For this, let's wait for the decision at the first PR.
> 
> * https://github.com/apache/spark/pull/33955/files#r708570515

Done

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 3]
```diff
@@ -3491,17 +3521,41 @@ case class ArrayDistinct(child: Expression)
             body
           }
 
-        val processArray = withArrayNullAssignment(
+        def withNaNCheck(body: String): String = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3524]: shall we move these codegen utils to `SQLOpenHashSet` as well to reduce duplicated code?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3524]: > shall we move these codegen utils to `SQLOpenHashSet` as well to reduce duplicated code?

How about to do this after all done.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3523]: @cloud-fan Updated, How about current?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala [hunk 4]
```diff
@@ -60,21 +60,52 @@ class SQLOpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
...
-        (value: Any) => java.lang.Float.isNaN(value.asInstanceOf[java.lang.Float])
-      case _ => (_: Any) => false
+        ((value: Any) => java.lang.Float.isNaN(value.asInstanceOf[java.lang.Float]),
+          java.lang.Float.NaN)
+      case _ => ((_: Any) => false, null)
     }
   }
 
-  def valueNaN(dataType: DataType): Any = {
+  def isNaNFuncAndValueNaN(dataType: DataType, valueName: String): Option[(String, String)] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala:75]: seems we can remove it.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala]: Yea, Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala [hunk 5]
```diff
@@ -60,21 +60,59 @@ class SQLOpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
...
-      case DoubleType => java.lang.Double.NaN
-      case FloatType => java.lang.Float.NaN
-      case _ => null
+  def withNaNCheckFunc(
+      isNaN: Any => Boolean,
+      valueNaN: Any,
+      value: Any,
+      hashSet: SQLOpenHashSet[Any],
+      handleNotNaN: () => Unit,
+      handleNaN: Any => Unit): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala:81]: can we make it return a function?
```
def withNaNCheckFunc(
    dataType: DataType,
    hashSet: SQLOpenHashSet[Any],
    handleNotNaN: Any => Unit,
    handleNaN: Any => Unit): Any => Unit = {
  (dataType match {
    case FloatType => ...
    case DoubleType => ...
  }).map { case (isNaN, valueNaN) =>
    (value: Any) => ...
  }.getOrElse {
    (value: Any) => handleNonNaN(value)
  }
}
```

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala]: Good suggestion. Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 6]
```diff
@@ -3410,32 +3410,60 @@ case class ArrayDistinct(child: Expression)
...
+    val data = array.asInstanceOf[ArrayData]
     doEvaluation(data)
   }
 
   @transient private lazy val doEvaluation = if (TypeUtils.typeWithProperEquals(elementType)) {
-    (data: Array[AnyRef]) => new GenericArrayData(data.distinct.asInstanceOf[Array[Any]])
+    (array: ArrayData) =>
+      val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+      val hs = new SQLOpenHashSet[Any]()
+      val (isNaN, valueNaN) = SQLOpenHashSet.isNaNFuncAndValueNaN(elementType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3421]: `isNaN` and `valueNaN` are only used by the next function call `withNaNCheckFunc`, we can calculate them in `withNaNCheckFunc`

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3421]: > `isNaN` and `valueNaN` are only used by the next function call `withNaNCheckFunc`, we can calculate them in `withNaNCheckFunc`

Yea, after  make it return a partial function, the value will be calculated once. Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 7]
```diff
@@ -3410,32 +3410,59 @@ case class ArrayDistinct(child: Expression)
...
+      val withNaNCheckFunc = SQLOpenHashSet.withNaNCheckFunc(elementType, hs,
+        (value: Any) =>
+          if (!hs.contains(value)) {
+            if (arrayBuffer.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+              ArrayBinaryLike.throwUnionLengthOverflowException(arrayBuffer.size)
+            }
+            arrayBuffer += value
+            hs.add(value)
+          },
+        (value: Any) => arrayBuffer += value)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3430]: nit: name it `valueNaN` to be clear

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Done

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2/3.1/3.0!

**Pull request closed without merging**